File: ports/go/src/core/kernel/bcsobjectcleanuphandler.go
package kernel

import "sync"

// BcsObjectCleanupHandler maps to QObjectCleanupHandler
type BcsObjectCleanupHandler struct {
	*BcsObject
	cleanupObjects []*BcsObject
	mu             sync.Mutex
}

func NewBcsObjectCleanupHandler() *BcsObjectCleanupHandler {
	return &BcsObjectCleanupHandler{
		BcsObject:      NewBcsObject(nil),
		cleanupObjects: make([]*BcsObject, 0),
	}
}

func (h *BcsObjectCleanupHandler) Add(obj *BcsObject) *BcsObject {
	h.mu.Lock()
	defer h.mu.Unlock()

	for _, o := range h.cleanupObjects {
		if o == obj {
			return obj
		}
	}
	h.cleanupObjects = append(h.cleanupObjects, obj)
	return obj
}

func (h *BcsObjectCleanupHandler) Remove(obj *BcsObject) {
	h.mu.Lock()
	defer h.mu.Unlock()

	for i, o := range h.cleanupObjects {
		if o == obj {
			h.cleanupObjects[i] = h.cleanupObjects[len(h.cleanupObjects)-1]
			h.cleanupObjects = h.cleanupObjects[:len(h.cleanupObjects)-1]
			return
		}
	}
}

func (h *BcsObjectCleanupHandler) IsEmpty() bool {
	h.mu.Lock()
	defer h.mu.Unlock()
	return len(h.cleanupObjects) == 0
}

func (h *BcsObjectCleanupHandler) Clear() {
	h.mu.Lock()
	defer h.mu.Unlock()
	for _, obj := range h.cleanupObjects {
		obj.Destroy()
	}
	h.cleanupObjects = nil
}

File: ports/rust/src/core/kernel/bcs_object_cleanup_handler.rs
use std::sync::{Arc, Mutex};
use super::bcs_object::BcsObject;

pub struct BcsObjectCleanupHandler {
    pub base: Arc<BcsObject>,
    cleanup_objects: Mutex<Vec<Arc<BcsObject>>>,
}

impl BcsObjectCleanupHandler {
    pub fn new() -> Arc<Self> {
        Arc::new(Self {
            base: BcsObject::new(),
            cleanup_objects: Mutex::new(Vec::new()),
        })
    }

    pub fn add(&self, obj: Arc<BcsObject>) -> Arc<BcsObject> {
        let mut objects = self.cleanup_objects.lock().unwrap();
        if !objects.iter().any(|o| Arc::ptr_eq(o, &obj)) {
            objects.push(obj.clone());
        }
        obj
    }

    pub fn remove(&self, obj: &Arc<BcsObject>) {
        let mut objects = self.cleanup_objects.lock().unwrap();
        if let Some(pos) = objects.iter().position(|o| Arc::ptr_eq(o, obj)) {
            objects.remove(pos);
        }
    }

    pub fn is_empty(&self) -> bool {
        self.cleanup_objects.lock().unwrap().is_empty()
    }

    pub fn clear(&self) {
        let objects = {
            let mut objs = self.cleanup_objects.lock().unwrap();
            let copy = objs.clone();
            objs.clear();
            copy
        };

        for obj in objects {
            obj.destroy();
        }
    }
}
